python/ray/tune/ray_trial_executor.py

```diff
@@ -949,28 +961,52 @@ def resource_string(self):
         else:
             return "? CPUs, ? GPUs"
 
-    def on_step_begin(self, trial_runner):
-        """Before step() called, update the available resources."""
+    def on_step_begin(
+            self,
+            trials: Union[List[Trial], "ray.tune.trial_runner.TrialRunner"]
+    ) -> None:
+        """Before step() is called, update the available resources."""
+        # avoid circular dependency
+        from ray.tune.trial_runner import TrialRunner
+        if isinstance(trials, TrialRunner):
+            logger.warning(
+                "TrialExecutor should not depend on TrialRunner interface. "
+                "Provide List[Trial] directly."
+            )
+            trials = trials.get_running_trials()
         self._update_avail_resources()
         self._trial_just_finished_before = self._trial_just_finished
         self._trial_just_finished = False
 
-    def on_step_end(self, trial_runner):
+    def on_step_end(
+            self,
+            trials: Union[List[Trial], "ray.tune.trial_runner.TrialRunner"]
+    ) -> None:
+        # avoid circular dependency
+        from ray.tune.trial_runner import TrialRunner
+        if isinstance(trials, TrialRunner):
+            logger.warning(
+                "TrialExecutor should not depend on TrialRunner interface. "
+                "Provide List[Trial] directly."
+            )
+            trials = trials.get_running_trials()
```

Comment:
quick question - shouldn't this check go into the TrialRunner?

We're putting the check in the implementation of the interface, but if another user implements the interface their code will just fail right?

Reply:
For on_step_end and all other similar interfaces, the change is non-breaking - accepting Union[List[Trial], TrialRunner]. The doc string of TrialExecutor.on_step_end is updated to nudge people to pass in List[Trial] directly (see file trial_executor.py).

The check here is only necessary if people may implement their own TrialRunner, which may in tern use RayTrialExecutor. I have no idea why/if that would happen but seems like TrialRunner is also in public folder? (correct me if this understanding is wrong.)

In terms of deprecating/privatizing/merging TrialExecutor and RayTrialExecutor, I will do that in a follow up PR. Basically if we detect a user somehow directly inherits from TrialExecutor, we print a warning msg that links them to the public RFC and asks them to contact us if any concern.

Reply:
Ah no one should be writing their own TrialRunner (there's no way to access it, and definitely private api).

The warning message sounds good!

Reply:
That's good to hear! Ok since we have full control of who calls RayTrialExecutor (basically TrainRunner that's also controlled by us), I removed these checks. I also enforced RayTrialExecutor to only take in List[Trial] but not TrianRunner. 

Note, current PR only gives user information in doc strings. I will soon put up another PR to inform user of deprecating this whole interface.

Reply:
great!

Reply:
Actually, I take it back :)
From what I heard from Kai, seems people may (already?) extend RayTrialExecutor?
If that's the case, I am afraid I have to make sure that RayTrialExecutor's API is also backward compatible.

Is that some flexibility that we want to provide to our user (extend RayTrialExecutor)?

Reply:
1. Let's look at how people are extending RayTrialExecutor. Ideally we understand how to capture their workloads in some better-defined interface (callback or something).
2. RayTrialExecutor API should be backwards compatible -- for now.